sonar-batch/src/main/java/org/sonar/batch/bootstrap/BatchWsClient.java

```diff
@@ -0,0 +1,130 @@
+
+  public BatchWsClient(WsClient target, boolean hasCredentials, @Nullable String publicBaseUrl) {
+    this.target = target;
+    this.hasCredentials = hasCredentials;
+    if (StringUtils.isBlank(publicBaseUrl)) {
+      this.publicBaseUrl = target.wsConnector().baseUrl();
+    } else {
+      this.publicBaseUrl = publicBaseUrl.replaceAll("(/)+$", "") + "/";
+    }
+  }
+
+  /**
+   * @throws IllegalStateException if the request could not be executed due to
+   *     a connectivity problem or timeout. Because networks can
+   *     fail during an exchange, it is possible that the remote server
+   *     accepted the request before the failure
+   * @throws HttpException if the response code is not in range [200..300)
+   */
+  public WsResponse call(WsRequest request) {
+    Profiler profiler = Profiler.createIfDebug(LOG).start();
+    WsResponse response = target.wsConnector().call(request);
+    profiler.stopDebug(format("%s %d %s", request.getMethod(), response.code(), response.requestUrl()));
+    failIfUnauthorized(response);
+    return response;
+  }
+
+  public String baseUrl() {
+    return target.wsConnector().baseUrl();
+  }
+
+  /**
+   * The public URL is optionally configured on server. If not, then the regular {@link #baseUrl()} is returned.
+   * URL has a trailing slash.
+   * See https://jira.sonarsource.com/browse/SONAR-4239
+   */
+  public String publicBaseUrl() {
+    return publicBaseUrl;
+  }
+
+  @VisibleForTesting
+  WsConnector wsConnector() {
+    return target.wsConnector();
+  }
+
+  private void failIfUnauthorized(WsResponse response) {
+    if (response.code() == HttpURLConnection.HTTP_UNAUTHORIZED) {
+      if (hasCredentials) {
+        // credentials are not valid
+        throw MessageException.of(format("Not authorized. Please check the properties %s and %s.",
+          CoreProperties.LOGIN, CoreProperties.PASSWORD));
+      }
+      // not authenticated - see https://jira.sonarsource.com/browse/SONAR-4048
+      throw MessageException.of(format("Not authorized. Analyzing this project requires to be authenticated. " +
+        "Please provide the values of the properties %s and %s.", CoreProperties.LOGIN, CoreProperties.PASSWORD));
+
+    }
+    if (response.code() == HttpURLConnection.HTTP_FORBIDDEN) {
+      // SONAR-4397 Details are in response content
+      throw MessageException.of(tryParseAsJsonError(response.content()));
+    }
+    response.failIfNotSuccessful();
+  }
+
+  private static String tryParseAsJsonError(String responseContent) {
+    try {
+      JsonParser parser = new JsonParser();
+      JsonObject obj = parser.parse(responseContent).getAsJsonObject();
+      JsonArray errors = obj.getAsJsonArray("errors");
+      List<String> errorMessages = new ArrayList<>();
+      for (JsonElement e : errors) {
+        errorMessages.add(e.getAsJsonObject().get("msg").getAsString());
+      }
+      return Joiner.on(", ").join(errorMessages);
+    } catch (Exception e) {
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) Either log or rethrow this exception. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1166)